src/MongoDB.Driver/Encryption/LibMongoCryptController.cs

```diff
@@ -0,0 +1,525 @@
+                }
+            }
+            catch (Exception ex)
+            {
+                throw new MongoClientException(ex.Message, ex);
+            }
+        }
+
+        public async Task<byte[]> EncryptFieldsAsync(string databaseName, byte[] unencryptedDocument, CancellationToken cancellationToken)
+        {
+            try
+            {
+                using (var context = GetCryptClient().StartEncryptionContext(databaseName, unencryptedDocument))
+                {
+                    return await ProcessStatesAsync(context, databaseName, cancellationToken).ConfigureAwait(false);
+                }
+            }
+            catch (Exception ex)
+            {
+                throw new MongoClientException(ex.Message, ex);
+            }
+        }
+
+        // private methods
+        private static bool AcceptAnyCertificate(
+            object sender,
+            X509Certificate certificate,
+            X509Chain chain,
+            SslPolicyErrors sslPolicyErrors)
+        {
+            return true;
+        }
+
+        private static IMongoCollection<BsonDocument> GetKeyVaultCollection(AutoEncryptionOptions autoEncryptionOptions, IMongoClient client)
+        {
+            var keyVaultClient = autoEncryptionOptions.KeyVaultClient ?? client;
+            var keyVaultNamespace = autoEncryptionOptions.KeyVaultNamespace;
+            var keyVaultDatabase = keyVaultClient.GetDatabase(keyVaultNamespace.DatabaseNamespace.DatabaseName);
+            return keyVaultDatabase.GetCollection<BsonDocument>(keyVaultNamespace.CollectionName);
+        }
+
+        private CryptClient GetCryptClient()
+        {
+            return _client.GetEncryptionClients().CryptClient;
+        }
+
+        private IMongoClient GetMongoCryptD()
+        {
+            // should not be called in constructor since it's a lazy
+            return _client.GetEncryptionClients().MongoCryptDClient;
+        }
+
+        private void FeedResult(CryptContext context, BsonDocument document)
+        {
+            var writerSettings = new BsonBinaryWriterSettings { GuidRepresentation = GuidRepresentation.Unspecified };
+            var documentBytes = document.ToBson(writerSettings: writerSettings);
+            context.Feed(documentBytes);
+            context.MarkDone();
+        }
+
+        private void FeedResults(CryptContext context, IEnumerable<BsonDocument> documents)
+        {
+            var writerSettings = new BsonBinaryWriterSettings { GuidRepresentation = GuidRepresentation.Unspecified };
+            foreach (var document in documents)
+            {
+                var documentBytes = document.ToBson(writerSettings: writerSettings);
+                context.Feed(documentBytes);
+            }
+            context.MarkDone();
+        }
+
+        private void ProcessErrorState(CryptContext context)
+        {
+            // todo: this stage doesn't work.
+            // Libmongocrypt throws an error together with stage.
```

Comment:
@rstam this stage probably is incorrect. During debugging sometimes I faced with `MONGOCRYPT_CTX_ERROR` stage. And this stage was at the same time with a thrown `libmongocrypt` exception.

Reply:
Should you implement this then?